File: android/app/src/main/kotlin/com/tensorroot/marketplace/MpTokenizerPlugin.kt
package com.tensorroot.marketplace

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.json.JSONObject

/**
 * Platform Channel plugin for Mercado Pago card tokenization.
 *
 * Uses Mercado Pago Core Methods SDK for PCI-compliant card tokenization.
 * Cards are never sent to our backend — only tokens.
 */
class MpTokenizerPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private var publicKey: String? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "initialize" -> {
                publicKey = call.argument<String>("publicKey")
                if (publicKey == null) {
                    result.error("INVALID_ARGS", "publicKey is required", null)
                    return
                }
                result.success(true)
            }

            "tokenizeCard" -> {
                val pk = publicKey
                if (pk == null) {
                    result.error("NOT_INITIALIZED", "Call initialize first", null)
                    return
                }

                val cardNumber = call.argument<String>("cardNumber")
                val expirationMonth = call.argument<String>("expirationMonth")
                val expirationYear = call.argument<String>("expirationYear")
                val securityCode = call.argument<String>("securityCode")
                val cardholderName = call.argument<String>("cardholderName")
                val identificationNumber = call.argument<String>("identificationNumber")
                val identificationType = call.argument<String>("identificationType") ?: "CPF"

                if (cardNumber == null || expirationMonth == null || expirationYear == null ||
                    securityCode == null || cardholderName == null || identificationNumber == null
                ) {
                    result.error("INVALID_ARGS", "All card fields are required", null)
                    return
                }

                tokenizeCard(
                    pk, cardNumber, expirationMonth, expirationYear,
                    securityCode, cardholderName, identificationNumber,
                    identificationType, result
                )
            }

            else -> result.notImplemented()
        }
    }

    private fun tokenizeCard(
        publicKey: String,
        cardNumber: String,
        expirationMonth: String,
        expirationYear: String,
        securityCode: String,
        cardholderName: String,
        identificationNumber: String,
        identificationType: String,
        result: MethodChannel.Result
    ) {
        // Use Mercado Pago REST API for tokenization (no SDK dependency needed)
        // This calls the public card token endpoint directly
        Thread {
            try {
                val url = java.net.URL("https://api.mercadopago.com/v1/card_tokens?public_key=$publicKey")
                val connection = url.openConnection() as java.net.HttpURLConnection
                connection.requestMethod = "POST"
                connection.setRequestProperty("Content-Type", "application/json")
                connection.doOutput = true

                val identificationJson = JSONObject().apply {
                    put("type", identificationType)
                    put("number", identificationNumber)
                }
                val cardholderJson = JSONObject().apply {
                    put("name", cardholderName)
                    put("identification", identificationJson)
                }
                val body = JSONObject().apply {
                    put("card_number", cardNumber)
                    put("expiration_month", expirationMonth)
                    put("expiration_year", expirationYear)
                    put("security_code", securityCode)
                    put("cardholder", cardholderJson)
                }.toString()

                connection.outputStream.use { it.write(body.toByteArray()) }

                val responseCode = connection.responseCode
                val responseBody = if (responseCode in 200..299) {
                    connection.inputStream.bufferedReader().readText()
                } else {
                    connection.errorStream?.bufferedReader()?.readText() ?: "Unknown error"
                }

                if (responseCode in 200..299) {
                    // Parse token response using JSONObject
                    val parsed = parseTokenResponse(responseBody)

                    android.os.Handler(android.os.Looper.getMainLooper()).post {
                        result.success(
                            mapOf(
                                "tokenId" to parsed["id"],
                                "lastFourDigits" to parsed["last_four_digits"],
                                "firstSixDigits" to parsed["first_six_digits"]
                            )
                        )
                    }
                } else {
                    android.os.Handler(android.os.Looper.getMainLooper()).post {
                        result.error("TOKENIZATION_ERROR", "Failed to tokenize card: $responseBody", null)
                    }
                }
            } catch (e: Exception) {
                android.os.Handler(android.os.Looper.getMainLooper()).post {
                    result.error("TOKENIZATION_ERROR", e.message, null)
                }
            }
        }.start()
    }

    private fun parseTokenResponse(responseBody: String): Map<String, String> {
        val json = JSONObject(responseBody)
        return mapOf(
            "id" to json.optString("id", ""),
            "first_six_digits" to json.optString("first_six_digits", ""),
            "last_four_digits" to json.optString("last_four_digits", ""),
            "expiration_month" to json.optString("expiration_month", ""),
            "expiration_year" to json.optString("expiration_year", ""),
        )
    }

    companion object {
        const val CHANNEL_NAME = "com.tensorroot.marketplace/mp_tokenizer"
    }
}

File: ios/Runner/MpTokenizerPlugin.swift
import Flutter
import UIKit

/// Platform Channel plugin for Mercado Pago card tokenization (iOS).
///
/// Calls the MP REST API /v1/card_tokens directly (PCI-compliant).
/// Cards are never sent to our backend — only tokens.
class MpTokenizerPlugin: NSObject, FlutterPlugin {
    static let channelName = "com.tensorroot.marketplace/mp_tokenizer"
    private var publicKey: String?

    static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: channelName,
            binaryMessenger: registrar.messenger()
        )
        let instance = MpTokenizerPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            guard let args = call.arguments as? [String: Any],
                  let pk = args["publicKey"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "publicKey is required", details: nil))
                return
            }
            publicKey = pk
            result(true)

        case "tokenizeCard":
            guard let pk = publicKey else {
                result(FlutterError(code: "NOT_INITIALIZED", message: "Call initialize first", details: nil))
                return
            }
            guard let args = call.arguments as? [String: Any],
                  let cardNumber = args["cardNumber"] as? String,
                  let expirationMonth = args["expirationMonth"] as? String,
                  let expirationYear = args["expirationYear"] as? String,
                  let securityCode = args["securityCode"] as? String,
                  let cardholderName = args["cardholderName"] as? String,
                  let identificationNumber = args["identificationNumber"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "All card fields are required", details: nil))
                return
            }
            let identificationType = (args["identificationType"] as? String) ?? "CPF"
            tokenizeCard(
                publicKey: pk,
                cardNumber: cardNumber,
                expirationMonth: expirationMonth,
                expirationYear: expirationYear,
                securityCode: securityCode,
                cardholderName: cardholderName,
                identificationNumber: identificationNumber,
                identificationType: identificationType,
                result: result
            )

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func tokenizeCard(
        publicKey: String,
        cardNumber: String,
        expirationMonth: String,
        expirationYear: String,
        securityCode: String,
        cardholderName: String,
        identificationNumber: String,
        identificationType: String,
        result: @escaping FlutterResult
    ) {
        guard let url = URL(string: "https://api.mercadopago.com/v1/card_tokens?public_key=\(publicKey)") else {
            result(FlutterError(code: "INVALID_URL", message: "Invalid MP API URL", details: nil))
            return
        }

        let body: [String: Any] = [
            "card_number": cardNumber,
            "expiration_month": expirationMonth,
            "expiration_year": expirationYear,
            "security_code": securityCode,
            "cardholder": [
                "name": cardholderName,
                "identification": [
                    "type": identificationType,
                    "number": identificationNumber
                ]
            ]
        ]

        guard let jsonData = try? JSONSerialization.data(withJSONObject: body) else {
            result(FlutterError(code: "SERIALIZATION_ERROR", message: "Failed to serialize request", details: nil))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = jsonData

        URLSession.shared.dataTask(with: request) { data, response, error in
            DispatchQueue.main.async {
                if let error = error {
                    result(FlutterError(code: "NETWORK_ERROR", message: error.localizedDescription, details: nil))
                    return
                }

                guard let data = data,
                      let httpResponse = response as? HTTPURLResponse else {
                    result(FlutterError(code: "TOKENIZATION_ERROR", message: "No response from MP API", details: nil))
                    return
                }

                guard (200...299).contains(httpResponse.statusCode) else {
                    let body = String(data: data, encoding: .utf8) ?? "Unknown error"
                    result(FlutterError(
                        code: "TOKENIZATION_ERROR",
                        message: "Failed to tokenize card: \(body)",
                        details: nil
                    ))
                    return
                }

                guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                      let tokenId = json["id"] as? String else {
                    result(FlutterError(code: "PARSE_ERROR", message: "Failed to parse MP token response", details: nil))
                    return
                }

                let lastFour = json["last_four_digits"] as? String
                let firstSix = json["first_six_digits"] as? String

                result([
                    "tokenId": tokenId,
                    "lastFourDigits": lastFour as Any,
                    "firstSixDigits": firstSix as Any
                ])
            }
        }.resume()
    }
}
